text/0000-pub-restricted.md

```diff
@@ -0,0 +1,924 @@
+   exposed outside of their module but still restrict the scope of
+   where they can be referenced, as discussed in the [motivation][]
+   section, and I do not think the current model can be "fixed" to
+   support that use case, at least not without adding some sort of
+   global reachability analysis as discussed in the previous bullet.
+
+In addition, these two alternatives do not address the main point
+being made in the [motivation][] section: one cannot tell exactly how
+"public" a `pub` item is, without working backwards through the module
+tree for all of its re-exports.
+
+## Curb your ambitions!
+
+ * Instead of adding support for restricting to arbitrary module
+   subtrees, narrow the feature to just `pub(crate) item`, so that one
+   chooses either "module private" (by adding no modifier), or
+   "universally visible" (by adding `pub`), or "visible to just the
+   current crate" (by adding `pub(crate)`).
+
+   This would be somewhat analogous to Java's relatively coarse
+   grained privacy rules, where one can choose `public`, `private`,
+   `protected`, or the unnamed "package" visibility.
+
+   I am all for keeping the implementation simple. However, the reason
+   that we should support arbitrary module subtrees is that doing so
+   will enable certain refactorings. Namely, if I decide I want to
+   inline the definition for one or more crates `A1`, `A2`, ... into
+   client crate `C` (i.e. replacing `extern crate A1;` with an
+   suitably defined `mod A1 { ... }`, but I do not want to worry about
+   whether doing so will risk future changes violating abstraction
+   boundaries that were previously being enforced via `pub(crate)`,
+   then I believe allowing `pub(path)` will allow a mechanical tool to
+   do the inline refactoring, rewriting each `pub(crate)` as `pub(A1)`
+   as necessary.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Can definition site fall outside restriction?
+[def-outside-restriction]: #can-definition-site-fall-outside-restriction
+
+For example, is it illegal to do the following:
+
+```rust
+mod a {
+  mod child { }
+  mod b { pub(super::child) const J: i32 = 3; }
+}
+```
+
+Or does it just mean that `J`, despite being defined in `mod b`, is
+itself not accessible in `mod b`?
+
+pnkfelix is personally inclined to make this sort of thing illegal,
+mainly because he finds it totally unintuitive, but is interested in
+hearing counter-arguments. Certainly the earlier [impl item example][]
+would look prettier as:
+
+```rust
+pub struct S;
+
+impl S {
+    pub(a) fn foo(&self) { println!("only callable within `a`"); }
+}
+
+mod a {
+    pub fn call_foo(s: &S) { s.foo(); }
+
+}
+
+fn rejected(s: &S) {
+    s.foo(); //~ ERROR: `S::foo` not visible outside of module `a`
+}
+```
+
+## Implicit Restriction Satisfaction (IRS:PUNPM)
+
+If a re-export occurs within a non-`pub` module, can we treat it as
+implicitly satisfying a restriction to `super` imposed by the item it
+is re-exporting?
```

Comment:
I'd prefer to start with the strict rules (i.e. the answer "No"). They fit better into the whole picture and can always be relaxed if necessary (I doubt it will be necessary).